Article Detail View with Comments

chrisgrande opened the pull request:
Features for #3 

- Live detail view
- Add comment field `turbo-frame` - _Shows as login button if not logged in_
- Deletable comments `turbo-frame` - _Checks if you own the comment_
- Editable comments `turbo-frame`-  _Checks if you own the comment_
- Lazy Load Comments via Turbo Beta 3

blopker commented on articles/views.py:
```diff
@@ -22,7 +56,11 @@ def get_success_url(self):
         """If the form is invalid, render the invalid form.
+        if form.is_valid():
+            new_comment = form.save(commit=False)
+            new_comment.body = form.cleaned_data['body']
```
Is thin intentional?

chrisgrande replied:
Not sure what you are asking.

blopker commented on articles/views.py:
```diff
@@ -1,15 +1,49 @@
+            new_comment.save()
+            return redirect('article_view', slug=slug)
+
+    return render(request, 'articles/_add_comment.html', {'form': form, 'article': article})
```
Dose this require the special redirect status codes?

chrisgrande replied:
I'm using _top on the frame currently to get it to work which causes the article page to just reload. Until we have streams to add the comments.

blopker commented on templates/articles/_add_comment.html:
```diff
@@ -0,0 +1,27 @@
+    </form>
+{% else %}
+    <div class="card m-5">
+        <button onclick="window.location.href='/login';" class="btn btn-sm btn-block btn-primary">Login to Comment</button>
```
Would be nice to use an actual link here for accessibility.

blopker replied:
You can also add a redirect query param so the user get put back here after login.

chrisgrande replied:
Yea, that was left over from a previous implementation. That doesn't even show up with the current way I have the `turbo-frame` loading it. If a user isn't logged in the the comment box just doesn't load.

blopker commented on templates/articles/_comment.html:
```diff
@@ -0,0 +1,27 @@
+        <a href="" class="comment-author">{{ comment.author }}</a>
+        <span class="date-posted">{{ comment.created_at }}</span>
+
+        {% if user.pk == comment.author.pk %}
```
I think the `author` is actually a `Profile`. So you may need to compare `user.profile == comment.author`?

chrisgrande replied:
Yea it is, odd that it worked my backwards way... corrected 👍🏻

blopker commented on templates/articles/_comment.html:
```diff
@@ -0,0 +1,27 @@
+            />
+        </a>
+        &nbsp;
+        <a href="" class="comment-author">{{ comment.author }}</a>
```
Does this work? If so, neat. I'd been using something like `comment.author.user.username`

luanfonceca replied:
I think that works because we have a `__str__` method in our Profile model:
![Screen Shot 2021-01-08 at 10 50 30](https://user-images.githubusercontent.com/1490875/104022664-5a34ed00-519f-11eb-80dc-3eec2dbf719a.png)


We also have a missing link:
```suggestion
        <a href="{% url 'profile_view' comment.author.user.username %}" class="comment-author">{{ comment.author }}</a>
```

We could implement the `get_absolute_url` to our Profile, avoiding a lot of those repetitions across our templates.

chrisgrande replied:
Yea I saw that as well - the profile also has a name method, but not of the generated users have first or last names so it returns an empty string, It should probably be updated to return first + last and if missing the username.

blopker commented on templates/articles/_comment.html:
```diff
@@ -0,0 +1,27 @@
+        {% endif %}
+    </div>
+</div>
+</turbo-frame>
```
Comments could be a fun spot to use Turbo Streams to get live updates! Not in this PR though :)

blopker reviewed (commented):
This is looking really good!

luanfonceca commented on templates/articles/_add_comment.html:
```diff
@@ -0,0 +1,27 @@
+                    src="{{ user.profile.img }}"
+                    class="comment-author-img"
+            />
+            <span class="comment-author">Commenting as <a href="">{{ user }}</a></span>
```
There's a missing link here:
```suggestion
            <span class="comment-author">Commenting as <a href="{% url 'profile_view' request.user.username %}">{{ user }}</a></span>
```

luanfonceca commented on templates/articles/_comment_confirm_delete.html:
```diff
@@ -0,0 +1,17 @@
+            <form method="post" action="{% url 'delete_comment' object.pk %}">
+                {% csrf_token %}
+                <input type="submit" class="btn btn-sm btn-danger" value="Confirm Delete">
+                <a href="{{ view.get_success_url }}" class="btn btn-default">Cancel</a>
```
I'm not sure if this is working, when I click the cancel button nothing happens on the interface.

In the network console, we have this call after the click:
![Screen Shot 2021-01-08 at 10 45 50](https://user-images.githubusercontent.com/1490875/104022165-af243380-519e-11eb-8b98-382f91d425bf.png)

chrisgrande replied:
Yea thats not working at the moment, I think it will need a view that can display the comment by itself.

chrisgrande replied:
Fixed 👍🏻

luanfonceca commented on templates/articles/_comment.html:
```diff
@@ -0,0 +1,27 @@
+        </p>
+    </div>
+    <div class="card-footer">
+        <a href="" class="comment-author">
```
Here's is missing the link as well:
```suggestion
        <a href="{% url 'profile_view' comment.author.user.username %}" class="comment-author">
```

luanfonceca commented on templates/articles/detail.html:
```diff
@@ -2,22 +2,22 @@
         <div class="info">
-          <a href="" class="author">Eric Simons</a>
-          <span class="date">January 20th</span>
+          <a href="" class="author">{{ article.author }}</a>
```
One more missing link:
```suggestion
          <a href="{% url 'profile_view' article.author.user.username %}" class="author">{{ article.author }}</a>
```

blopker commented:
Hey @chrisgrande, can you drop a comment when you're ready for another review? Thanks!

chrisgrande commented:
@blopker — I think this is ready to go.

blopker commented on templates/articles/_comment.html:
```diff
@@ -0,0 +1,34 @@
+{% if view.comment_deleted %}
+    <turbo-frame id="comment-{{ view.kwargs.pk }}">
+     <!-- Comment Deleted -->
+    </turbo-frame>
```
Should this be a stream response? Like:

```
<turbo-stream action="remove" target="comment-{{ view.kwargs.pk }}">
</turbo-stream>
```

chrisgrande replied:
Ohh, I totally forgot that was a thing without websockets. Let me give that a go.

chrisgrande replied:
That works much better than my blank trick. I just had to set the header on the view. That can be swapped out when we have a definitive set of mixins, similar to: https://github.com/hotwire-django/django-turbo-response/blob/main/src/turbo_response/response.py

blopker commented:
It works really well! Thanks @chrisgrande. This can definitely get merged now. I just wonder if we could to with with (non-websocket) streams without having to do that blank turbo-frame trick to delete?

blopker commented on articles/views.py:
```diff
@@ -1,16 +1,75 @@
+        return reverse_lazy("login_to_add_comment", kwargs={"slug": self.kwargs["slug"]})
+
+    def get_success_url(self):
+        return reverse_lazy("article_view", kwargs={"slug": self.kwargs["slug"]})
```
Instead of redirecting, can we also return a `prepend` stream response? That way we don't have to do a page refresh and the scroll position isn't lost.

chrisgrande replied:
Yea, I was just looking at that too. I was wondering how much we want to do before we have a package to add and extract things with.

blopker replied:
Our main purpose here is to find patterns in real apps to abstract with a package. We'd want to make all features in this app as idiomatic as possible so we can do that abstraction later once we see what the patterns are. I think this feature in particular is one that shows up a lot and is a great example to explore the ideal implementation with.

chrisgrande replied:
Good point, it's certainty a scenario that appears regularly.

chrisgrande committed:
Fix #3: Detail View with Comments

Base Detail view wiring

Favorite and Follow Counts

Display comments

Comment code reformat

Profile image fix and move add comment to partial

Login comment button - body line breaks

Add comment profile picture and name

Ability to add comments

Quick implementation using a turbo-frame

Delete comment

Cleanup the confirm partial

Consolidate model imports

Suggestions from Pull Request

Co-authored-by: Luan Fonseca <luanfonceca@gmail.com>

Update templates/articles/_comment.html

Co-authored-by: Luan Fonseca <luanfonceca@gmail.com>

Update templates/articles/detail.html

Co-authored-by: Luan Fonseca <luanfonceca@gmail.com>

Update templates/articles/_comment.html

Co-authored-by: Luan Fonseca <luanfonceca@gmail.com>

Compare profiles

Move to CBV

Redirect comment login to article

Remove comment on delete

Delete comment notes

Mixin for UserIsAuthor

Fix ViewComment Object Type

Update Comment

Rename Comment Form

Remove Comment with turbo-stream

Remove TODO

Prepend comments with turbo-stream

Fix “Login to Comment” full page navigation

Fix Views

Missing Import

chrisgrande commented:
Rebased onto Fix #22. @blopker I think it's ready to go, what do you think?

chrisgrande committed:
Update to Turbo Beta 3

Fixes for MIME type change

chrisgrande committed:
Profile links fix to break out of turbo-frame

chrisgrande committed:
Lazy Load Article Comments
